core/src/banking_stage/forwarder.rs

```diff
@@ -175,13 +175,9 @@ impl Forwarder {
         });
 
         let packet_vec: Vec<_> = forwardable_packets
-            .filter_map(|p| {
-                if !p.meta().forwarded() && self.data_budget.take(p.meta().size) {
-                    Some(p.data(..)?.to_vec())
-                } else {
-                    None
-                }
-            })
+            .filter(|p| !p.meta().forwarded())
```

Comment:
The new code is cleaner definitely. Isn't the left side more efficient though without multiple loops?

Reply:
It won't do multiple loops, iterators are a zero-cost abstraction: https://doc.rust-lang.org/book/ch13-04-performance.html

> Closures and iterators are Rust features inspired by functional programming language ideas. They contribute to Rust’s capability to clearly express high-level ideas at low-level performance. The implementations of closures and iterators are such that runtime performance is not affected. This is part of Rust’s goal to strive to provide zero-cost abstractions.

They get compiled directly, so expansion isn't really a thing, but we kind of think how expanding these two versions might work...
*Roughly* the code before was:
```rust
    let mut packet_vec = vec![];
    for p in forwardable_packets {
        if !p.meta().forwarded() && self.data_budget.take(p.meta().size) {
            if let Some(data) = p.data(..) {
                packet_vec.push(data.to_vec());
            }
        }
    }
```

And after the change, (again roughly):
```rust
    let mut packet_vec = vec![];
    for p in forwardable_packets {
        if p.meta().forwarded() { continue; }
        if !p.data_budget.take(p.meta().size) { continue; }
        if let Some(data) = p.data(..) {
            packet_vec.push(data.to_vec());
        }
    }
```

Reply:
Good to know! Thanks